staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go

```diff
@@ -1005,3 +1005,88 @@ const (
+	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,1,opt,name=apiVersion"`
+	// Fields TODO
+	Fields FieldSet `json:"fields,omitempty" protobuf:"bytes,2,opt,name=fields,casttype=FieldSet"`
+}
+
+// FieldSet identifies a set of fields.
+type FieldSet struct {
+	// Members lists fields that are part of the set.
+	Members []FieldPathElement `json:"members" protobuf:"bytes,1,rep,name=members"`
+	// Children lists child fields which themselves have children that are
+	// members of the set. Appearance in this list does not imply membership.
+	// Note: this is a tree, not an arbitrary graph.
+	Children []FieldSetNode `json:"children" protobuf:"bytes,2,rep,name=children"`
+}
+
+// FieldPathElement describes how to select a child field given a containing object.
+type FieldPathElement struct {
+	// Exactly one of the following fields should be non-nil.
+	// FieldName selects a single field from a map (reminder: this is also
+	// how structs are represented). The containing object must be a map.
+	// +optional
+	FieldName *string `json:"fieldName,omitempty" protobuf:"bytes,1,opt,name=fieldName"`
+	// Key selects the list element which has fields matching those given.
+	// The containing object must be an associative list with map typed
+	// elements.
+	// +optional
+	Key []FieldValuePair `json:"key,omitempty" protobuf:"bytes,2,rep,name=key"`
+	// Value selects the list element with the given value. The containing
+	// object must be an associative list with a primitive typed element
+	// (i.e., a set).
+	// +optional
+	Value *Value `json:"value,omitempty" protobuf:"bytes,3,opt,name=value"`
+	// Index selects a list element by its index number. The containing
+	// object must be an atomic list.
+	// +optional
+	Index *int32 `json:"index,omitempty" protobuf:"varint,4,opt,name=index"`
+}
+
+// FieldSetNode is a pair of FieldPathElement / FieldSet, for the purpose of expressing
+// nested set membership.
+type FieldSetNode struct {
+	// PathElement TODO
+	PathElement FieldPathElement `json:"pathElement" protobuf:"bytes,1,opt,name=pathElement"`
+	// Set TODO
+	Set *FieldSet `json:"set" protobuf:"bytes,2,opt,name=set"`
+}
+
+// FieldValuePair is an individual key-value pair.
+type FieldValuePair struct {
+	// Name TODO
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// Value TODO
+	Value Value `json:"value" protobuf:"bytes,2,opt,name=value"`
+}
+
+// A Value is an object; it corresponds to an 'atom' in the schema.
+type Value struct {
+	// Exactly one of the following fields should be set.
+	// Float TODO
+	// +optional
+	Float *float64 `json:"float,omitempty" protobuf:"bytes,1,opt,name=float"`
+	// Int TODO
+	// +optional
+	Int *int32 `json:"int,omitempty" protobuf:"varint,2,opt,name=int"`
+	// String TODO
+	// +optional
+	Str *string `json:"str,omitempty" protobuf:"bytes,3,opt,name=str"`
+	// Boolean TODO
+	// +optional
+	Boolean *bool `json:"boolean,omitempty" protobuf:"varint,4,opt,name=boolean"`
+	// List is a list of items.
```

Comment:
I think we can omit these from here down. We don't need to support e.g. atomic lists as keys in the first take, or maybe ever?

Reply:
Further thought-- what if we just used a string for every value? It's easy to parse into a number if needed and it would simplify this API a lot?

Reply:
So, more like unstructured? That sounds good but then could we still add support for  atomic lists as keys if we needed to later?

Reply:
Yes, as long as they can be serialized. We could alternatively keep this struct and have it be a union of only a single field, so we could extend in the future if we wanted.

Reply:
This would also disable having associative lists of atomic lists, is that okay?

Reply:
What about something like this: https://github.com/kubernetes/kubernetes/blob/5a16163c87fe2a90916a51b52771a668bcaf2a0d/pkg/apis/admissionregistration/types.go#L202

Atomic?

Reply:
Yes I think that's atomic (doesn't make sense for different parts to have different owners).